File: day-7/main.py
class node:
    def __init__(self, parent=None):
        self.parent = parent
        self.fileList = []
        self.size = 0
        self.children = {}

class file:
    def __init__(self, name, size):
        self.size = size
        self.name = name

def readInput(fileName):
    return([x.strip() for x in open(fileName, 'r').readlines()])

class command:
    def __init__(self, command, argument=None):
        if(command == 'c'):
            self.command = 'cd'
            self.argument = argument
        elif(command == 'l'):
            self.command = 'ls'
            self.argument = None

def parseInstruction(input):
    if input[0] == '$':
        if input[2] == 'c':
            return command('c', input[5:])
        elif input[2] == 'l':
            return command('l')

def parseInput(input, root):
    wd = root
    ls = False
    for instruction in input:
        print(instruction)
        if instruction[0] == '$':
            ls = False
        if not ls:
            instruction = parseInstruction(instruction)
            if isinstance(instruction, command):
                if(instruction.command == 'ls'):
                    ls = True
                    continue
                elif(instruction.command == 'cd'):
                    ls = False
                    if(instruction.argument == '/'):
                        wd = root
                    elif(instruction.argument == '..'):
                        if(wd != root):
                            wd = wd.parent
                    else:
                        print(isinstance(wd.children[instruction.argument], node))
                        wd = wd.children[instruction.argument]
        else:
            if(instruction[:3] == 'dir'):
                directory = node(wd)
                wd.children[instruction[4:]] = [directory]
            else:
                print(instruction)
                name, size = instruction.split(" ")
                print(type(wd))
                print(dir(wd))
                print(isinstance(wd, node))
                wd.fileList.append(file(name, size))


def main():
    input = readInput('input')
    root = node(None)
    parseInput(input, root)

main()

File: day-7/src/main.rs
#[allow(unused_imports)]
use std::collections::HashMap;
use std::fs;

#[allow(dead_code)]
struct File {
    name: String,
    size: usize,
}

impl File {
    pub fn new(name: String, size: usize) -> Self {
        Self { name, size }
    }
}
#[allow(dead_code)]
struct Directory {
    name: String,
    parent: Option<Box<Directory>>,
    file_list: Vec<File>,
    size: usize,
    children: Vec<Directory>,
}
#[allow(dead_code)]
impl Directory {
    pub fn new(parent: Option<Box<Directory>>, name: String) -> Self {
        Self {
            name,
            parent,
            file_list: Vec::new(),
            size: 0,
            children: Vec::new(),
        }
    }
}
#[allow(dead_code)]
enum Command {
    Cd { argument: String },
    Ls,
}
#[allow(dead_code)]
enum CommandOrFileOrDir {
    Command(Command),
    File(File),
    Node(Directory),
}

fn read_input(path: String) -> Vec<String> {
    let contents = fs::read_to_string(path).expect("should have been able to read the file");
    let contents: Vec<String> = contents.lines().map(|s| s.to_owned()).collect();
    contents
}
#[allow(dead_code)]
fn parse_line(
    line: String,
    wd: Option<Box<Directory>>,
) -> Result<CommandOrFileOrDir, &'static str> {
    if line.as_bytes()[0] as char == '$' {
        if line.as_bytes()[2] as char == 'l' {
            Ok(CommandOrFileOrDir::Command(Command::Ls))
        } else {
            Ok(CommandOrFileOrDir::Command(Command::Cd {
                argument: line[5..].to_string(),
            }))
        }
    } else if line.as_bytes()[0] as char == 'd' {
        Ok(CommandOrFileOrDir::Node(Directory::new(
            wd,
            line[4..].to_string(),
        )))
    } else {
        let size: usize = (line.split(' ').next()).unwrap().parse().unwrap();
        let name = line.split(' ').nth(1);
        Ok(CommandOrFileOrDir::File(File::new(
            name.unwrap().to_string(),
            size,
        )))
    }
}

fn main() {
    let contents = read_input("input".to_owned());
    for line in contents {
        println!("{}", line.as_bytes()[0] as char);
        println!("{}", line);
    }
}
